blockchain/chain_sync/src/network_context.rs

```diff
@@ -62,6 +64,45 @@ impl<DB: Clone> Clone for SyncNetworkContext<DB> {
     }
 }
 
+struct RaceBatch<T> {
+    tasks: JoinSet<Result<T, String>>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl<T> RaceBatch<T>
+where
+    T: Send + 'static,
+{
+    pub fn new(max_concurrent_jobs: usize) -> Self {
+        RaceBatch {
+            tasks: JoinSet::new(),
+            semaphore: Arc::new(Semaphore::new(max_concurrent_jobs)),
+        }
+    }
+
+    pub fn add(&mut self, future: impl Future<Output = Result<T, String>> + Send + 'static) {
+        let sem = self.semaphore.clone();
+        self.tasks.spawn(async move {
+            let permit = sem.acquire_owned().await.unwrap();
```

Comment:
Unwraps make me nervous.

```suggestion
            let permit = sem.acquire_owned().await.map_err(|_| "Semaphore unexpectedly closed")?;
```